servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/AbstractH2DuplexHandler.java

```diff
@@ -76,13 +75,10 @@ final void writeMetaData(ChannelHandlerContext ctx, HttpMetaData metaData, Http2
         ctx.write(new DefaultHttp2HeadersFrame(h2Headers, endStream), promise);
     }
 
-    static void writeBuffer(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        Buffer buffer = (Buffer) msg;
-        if (buffer.readableBytes() > 0) {
-            ctx.write(new DefaultHttp2DataFrame(encodeAndRetain(buffer), false), promise);
-        } else {
-            promise.setSuccess();
-        }
+    static void writeBuffer(ChannelHandlerContext ctx, Buffer buffer, ChannelPromise promise) {
+        // Even if the Buffer is empty we still want to write it. This is to preserve order of notification for
+        // when writing empty payload in order to know when prior content is written (and then close the channel).
+        ctx.write(new DefaultHttp2DataFrame(encodeAndRetain(buffer), false), promise);
```

Comment:
original discussion: https://github.com/apple/servicetalk/pull/1850#discussion_r717129692

@idelpivnitskiy - this was done so we can preserve order of promise completion relative to writes. in some cases we may not have a reference to the "last write" and may do something like `write(empty buffer).whenComplete(close)`.

looks like netty shouldn't [write empty data frames to the wire](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java#L484-L488), was there a scenario that lead to writing the empty data frames to the wire?

I saw in a commit message related to this change mention of a race between cancel and write on the write stream subscriber, but you made a recent fix which seems related https://github.com/apple/servicetalk/commit/4cb71f59afecc3c1cd1d6fc0f050cd98c0134eb9 ... let me know if you still feel we have a race with this approach as that maybe concerning in other scenarios.

Reply:
Thanks for clarifying. Consider creating a static constant as `new DefaultHttp2DataFrame(false)` to avoid allocation on empty buffers.
We can also use a constant for `new DefaultHttp2DataFrame(true)` in line 100.

Reply:
I considered this but `DefaultHttp2DataFrame` has state (`Http2FrameStream` state & id) that Netty uses internally.

Reply:
True, forgot to check that `AbstractHttp2StreamFrame` has a non-final field.